lib/sqlalchemy/sql/type_api.py

```diff
@@ -470,13 +470,20 @@ def _generic_type_affinity(self):
                     "sqlalchemy.sql.sqltypes",
                     "sqlalchemy.sql.type_api",
                 )
-                and t._is_generic_type()
+                and issubclass(t, TypeEngine)
+                and t is not TypeEngine
             ):
-                if t in (TypeEngine, UserDefinedType):
-                    return NULLTYPE.__class__
-                return t
-        else:
-            return self.__class__
+                if t.__name__.isupper() and not best_uppercase:
+                    best_uppercase = t
+                elif not t.__name__.isupper() and not best_camelcase:
+                    best_camelcase = t
+
+        return best_camelcase or best_uppercase or NULLTYPE.__class__
+
+    def as_generic(self):
+        """Return an instance of the generic type corresponding to this type"""
```

Comment:
**mike bayer** (zzzeek) wrote:

I would add an example in doctest style:

    e.g.::

        >>> from sqlalchemy.dialects.mysql import INTEGER
        >>> INTEGER(display_width=4).as_generic()
        Integer()

then we also add:

    .. versionadded:: 1.4.0b2


is this OK for 1.4?  I think you have a workaround for 1.3.x series now right?

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2375